builtins: Implement ST_Node function

mknycha opened the pull request:
This function adds a node on a geometry for each intersection

Release note (sql change): Implement geo builtin ST_Node

Resolves #48993

This PR is dependent on additional package - introduced here: https://github.com/cockroachdb/vendored/pull/39

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56183)
<!-- Reviewable:end -->

otan commented on go.mod:
```diff
@@ -128,6 +128,7 @@ require (
 	github.com/pierrec/lz4 v2.4.1+incompatible // indirect
 	github.com/pierrre/geohash v1.0.0
 	github.com/pkg/browser v0.0.0-20180916011732-0a3d74bf9ce4
+	github.com/pkg/errors v0.9.1
```
please revert this

mknycha replied:
ok, done

otan commented on pkg/geo/geomfn/node.go:
```diff
@@ -0,0 +1,124 @@
+	"github.com/twpayne/go-geom/xy"
+)
+
+// extractUniqueEndpoints extracts the endpoints from geometry and removes duplicates
```
nit: `.` at end of sentence.

mknycha replied:
done

otan commented on pkg/geo/geomfn/node.go:
```diff
@@ -0,0 +1,124 @@
+	return output, nil
+}
+
+// extractEndpoints extracts the endpoints from geometry provided and returns them as a multipoint geometry
```
nit: `.` at end of sentence.
spell multipoint as MultiPoint

mknycha replied:
done

otan commented on pkg/geo/geomfn/node.go:
```diff
@@ -0,0 +1,124 @@
+func extractUniqueEndpoints(g geo.Geometry) (geo.Geometry, error) {
+	endpoints, err := extractEndpoints(g)
+	if err != nil {
+		return geo.Geometry{}, fmt.Errorf("error extracting endpoints: %v", err)
```
use errors.Newf

mknycha replied:
done

otan commented on pkg/geo/geomfn/node.go:
```diff
@@ -0,0 +1,124 @@
+
+	output, err := UnaryUnion(endpoints)
+	if err != nil {
+		return geo.Geometry{}, fmt.Errorf("error removing duplicates: %v", err)
```
use errors.Newf

mknycha replied:
done

otan commented on pkg/geo/geomfn/node.go:
```diff
@@ -0,0 +1,124 @@
+	if !xy.IsOnLine(l.Layout(), p, l.FlatCoords()) {
+		return false, []*geom.LineString{l}, nil
+	}
+	// Do not split if point is the endpoint of the line
```
nit: period at end of comment

mknycha replied:
done

otan commented on pkg/geo/geomfn/node.go:
```diff
@@ -0,0 +1,124 @@
+	coordsA := []geom.Coord{}
+	coordsB := []geom.Coord{}
+	for i := 1; i < l.NumCoords(); i++ {
+		if xy.IsPointWithinLineBounds(p, l.Coord(i-1), l.Coord(i)) {
```
nit: period at end of comment

mknycha replied:
done

otan commented on pkg/geo/geomfn/topology_operations.go:
```diff
@@ -188,3 +193,75 @@ func SharedPaths(a geo.Geometry, b geo.Geometry) (geo.Geometry, error) {
+		return geo.Geometry{}, err
+	}
+	if lines.ShapeType() == geopb.ShapeType_LineString {
+		// No nodes found, return a multilinestring
```
nit: call this MultiLineString, `.` at end of comment

mknycha replied:
done

otan commented on pkg/geo/geomfn/topology_operations.go:
```diff
@@ -188,3 +193,75 @@ func SharedPaths(a geo.Geometry, b geo.Geometry) (geo.Geometry, error) {
 	return gm, nil
 }
+
+// Node Returns a geometry containing a set of linestrings using the least possible number of nodes while preserving all of the input ones.
```
nit: let's break this up:

```
// Node returns a geometry containing a set of LineStrings using the least
// possible number of nodes while preserving all of the input ones.

```

Also, feel free to move this function and it's tests to node.go.

mknycha replied:
that's a good idea 👍

otan commented on pkg/geo/geomfn/topology_operations.go:
```diff
@@ -188,3 +193,75 @@ func SharedPaths(a geo.Geometry, b geo.Geometry) (geo.Geometry, error) {
+	}
+	ep, err := extractUniqueEndpoints(g)
+	if err != nil {
+		return geo.Geometry{}, fmt.Errorf("error extracting unique endpoints: %v", err)
```
use errors.Newf

mknycha replied:
done

otan commented on pkg/geo/geomfn/topology_operations.go:
```diff
@@ -188,3 +193,75 @@ func SharedPaths(a geo.Geometry, b geo.Geometry) (geo.Geometry, error) {
+	if g.ShapeType() != geopb.ShapeType_LineString && g.ShapeType() != geopb.ShapeType_MultiLineString {
+		return geo.Geometry{}, errors.New("geometry type is unsupported. Please pass a LineString or a MultiLineString")
+	}
+	ep, err := extractUniqueEndpoints(g)
```
nit: move this down closer to where it's actually used

mknycha replied:
done

otan commented on pkg/geo/geomfn/topology_operations.go:
```diff
@@ -188,3 +193,75 @@ func SharedPaths(a geo.Geometry, b geo.Geometry) (geo.Geometry, error) {
+	}
+	mllines := glines.(*geom.MultiLineString)
+
+	mlout := geom.NewMultiLineString(geom.XY)
```
can you add a comment about what this section of code is doing?

mknycha replied:
I will move it to a separate function with a comment

otan commented on pkg/geo/geomfn/node.go:
```diff
@@ -0,0 +1,124 @@
+	return nil
+}
+
+// splitLineByPoint splits the line using the point provided. Returns true and a slice of output line strings.
```
nit: instead of "Returns true and", it should be "returns a bool representing ...."

mknycha replied:
fixed

otan commented on pkg/geo/geomfn/node.go:
```diff
@@ -0,0 +1,124 @@
+		return geo.Geometry{}, fmt.Errorf("error extracting endpoints: %v", err)
+	}
+
+	output, err := UnaryUnion(endpoints)
```
why do we call UnaryUnion?

mknycha replied:
It was supposed to remove duplicated points. I have just realised that it's not needed.

otan replied:
i think it still is.

mknycha replied:
I mean, later in the code, it's going through each endpoint for each line, it breaks the loop once it finds any point for a particular line - this way removing `UnaryUnion` here should not affect the results, but it may affect the performance. I have added it here because it's used in PostGIS ([link](http://postgis.net/docs/doxygen/2.4/dd/d8f/lwgeom__geos__node_8c_a0ae5516a2dcf5d724c4b1652433076e5.html#a0ae5516a2dcf5d724c4b1652433076e5)). Should I keep it?

otan replied:
If we check it later already then let's not do it here. Is there a test for this case?

mknycha replied:
Yes, there is https://github.com/cockroachdb/cockroach/pull/56183/files#diff-f38103c4687b9c33fd756280757bbcafc674606264c9b5fdabc6b48b2868a7d6R484-R489, but for a duplicated LineString. Let me add one more for a LineString with duplicated points.

otan commented on pkg/geo/geomfn/node.go:
```diff
@@ -0,0 +1,124 @@
+	return result, nil
+}
+
+// collectEndpoints takes endpoints of the line provided and adds them to the provided multipoint
```
nit: period at end of sentence.

I think this function should return a list of `geom.Coord`, and the calling function should add it to the existing MultiPoint collection to make this function purer with no side effects.

mknycha replied:
Makes sense, but should not it return a list of `geom.Point`s?

otan replied:
either works in this case.

otan reviewed (commented):
Hey, thanks for doing this!

This is mostly nits about commenting. Also, all fmt.Errorf should be errors.Newf.

mknycha committed:
builtins: Implement ST_Node function
This function adds a node on a geometry for each intersection

Release note (sql change): Implement geo builtin ST_Node

mknycha commented:
@otan it's ready to be reviewed again

otan commented:
Currently on leave so I'll look when I'm back next week!

On Thu, Nov 5, 2020 at 1:29 PM Marcin Knychała <notifications@github.com>
wrote:

> @otan <https://github.com/otan> it's ready to be reviewed again
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/56183#issuecomment-722656592>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA32FQ53POFL23IZQBM4LPTSOMKLHANCNFSM4TF6Y4IA>
> .
>


-- 
Oliver Tan | Member of Technical Staff
Cockroach Labs

otan reviewed (approved):
cool, thanks!

bors r+

mknycha commented:
Thanks for the review, sorry for bothering you during your holiday. I'll be more patient next time :)

otan commented:
no worries -- thanks for your contribution!!!
